File: 242_valid_anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function (s, t) {
  var counts = {};

  if (s.length !== t.length) {
    return false;
  }

  for (var i = 0; i < s.length; i++) {
    if (counts[s[i]]) {
      counts[s[i]]++;
    } else {
      counts[s[i]] = 1;
    }

    if (counts[t[i]]) {
      counts[t[i]]--;
    } else {
      counts[t[i]] = -1;
    }
  }

  for (var key in counts) {
    if (counts[key] !== 0) {
      return false;
    }
  }

  return true;
};

File: 242_valid_anagram.rb
# @param {String} s
# @param {String} t
# @return {Boolean}
def is_anagram(s, t)
  counts = Hash.new(0)
  
  s.each_char { |char| counts[char] += 1 }
  t.each_char { |char| counts[char] -= 1 }
  
  counts.all? { |k, v| v == 0 }
end